File: examples/canonical-auth/java/src/main/java/dev/tenantcore/auth/routes/UserController.java
package dev.tenantcore.auth.routes;

import dev.tenantcore.auth.auth.AuthContext;
import dev.tenantcore.auth.domain.InMemoryStore;
import dev.tenantcore.auth.domain.User;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.Collection;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/users")
public class UserController {

    private final InMemoryStore store;

    public UserController(InMemoryStore store) {
        this.store = store;
    }

    private AuthContext getAuth() {
        return (AuthContext) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
    }

    @GetMapping
    @PreAuthorize("hasAuthority('iam:user:read')")
    public ResponseEntity<Collection<User>> listUsers() {
        return ResponseEntity.ok(store.getAllUsers(getAuth().tenantId()));
    }

    @PostMapping
    @PreAuthorize("hasAuthority('iam:user:invite')")
    public ResponseEntity<?> inviteUser(@RequestBody Map<String, Object> body) {
        User user = store.inviteUser(
            getAuth().tenantId(),
            String.valueOf(body.get("email")),
            String.valueOf(body.getOrDefault("display_name", body.getOrDefault("displayName", "Invited User"))),
            body.get("group_slugs") instanceof List<?> groupSlugs
                ? groupSlugs.stream().filter(String.class::isInstance).map(String.class::cast).toList()
                : List.of()
        );
        return ResponseEntity.status(HttpStatus.CREATED).body(user);
    }

    @GetMapping("/{id}")
    @PreAuthorize("hasAuthority('iam:user:read')")
    public ResponseEntity<?> getUser(@PathVariable String id) {
        User user = store.getUserById(id);
        if (user == null || !java.util.Objects.equals(user.tenantId(), getAuth().tenantId())) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        return ResponseEntity.ok(user);
    }

    @PatchMapping("/{id}")
    @PreAuthorize("hasAuthority('iam:user:update')")
    public ResponseEntity<?> updateUser(@PathVariable String id, @RequestBody Map<String, Object> body) {
        User existing = store.getUserById(id);
        if (existing == null || !java.util.Objects.equals(existing.tenantId(), getAuth().tenantId())) {
            return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
        }
        return ResponseEntity.ok(Map.of(
            "id", existing.id(),
            "email", existing.email(),
            "display_name", body.getOrDefault("display_name", existing.displayName())
        ));
    }
}

File: examples/canonical-auth/elixir/lib/tenantcore_auth/controllers/user_controller.ex
defmodule TenantcoreAuth.UserController do
  use TenantcoreAuth, :controller

  alias TenantcoreAuth.Domain.InMemoryStore

  plug TenantcoreAuth.Auth.RbacPlug, "iam:user:read" when action in [:index, :show]
  plug TenantcoreAuth.Auth.RbacPlug, "iam:user:invite" when action in [:create]
  plug TenantcoreAuth.Auth.RbacPlug, "iam:user:update" when action in [:update]

  def index(%Plug.Conn{assigns: %{auth: auth}} = conn, _params) do
    users =
      InMemoryStore.list_users(auth.tenant_id)
      |> Enum.map(&serialize_user/1)

    json(conn, %{users: users})
  end

  def show(%Plug.Conn{assigns: %{auth: auth}} = conn, %{"id" => id}) do
    case InMemoryStore.get_user_by_id(id) do
      %{tenant_id: tenant_id} = user when tenant_id == auth.tenant_id ->
        json(conn, %{user: serialize_user(user)})

      _ ->
        conn |> put_status(:forbidden) |> json(%{error: "Forbidden"})
    end
  end

  def create(%Plug.Conn{assigns: %{auth: auth}} = conn, %{"email" => email, "display_name" => display_name} = params) do
    user = InMemoryStore.invite_user(auth.tenant_id, email, display_name, Map.get(params, "group_slugs", []))
    conn |> put_status(:created) |> json(%{user: serialize_user(user)})
  end

  def update(%Plug.Conn{assigns: %{auth: auth}} = conn, %{"id" => id, "display_name" => display_name}) do
    case InMemoryStore.get_user_by_id(id) do
      %{tenant_id: tenant_id} = user when tenant_id == auth.tenant_id ->
        user = %{user | display_name: display_name}
        json(conn, %{user: serialize_user(user)})

      _ ->
        conn |> put_status(:forbidden) |> json(%{error: "Forbidden"})
    end
  end

  defp serialize_user(user) do
    %{id: user.id, email: user.email, display_name: user.display_name, tenant_id: user.tenant_id, acl_ver: user.acl_ver}
  end
end
